src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -0,0 +1,402 @@
+        {
+            FileEncoding encodingPreference = FileEncoding.Unspecified;
+            try
+            {
+                // It doesn't matter if this fails or throws, we will return unknown in that case
+                object tmp = sessionState.PSVariable.GetValue("PSDefaultFileEncoding");
+                LanguagePrimitives.TryConvertTo<FileEncoding>(tmp, out encodingPreference);
+            }
+            catch
+            {
+                ;
+            }
+            return encodingPreference;
+        }
+
+        /// <summary>
+        /// Retrieve the encoding in a provider context
+        /// </summary>
+        public static Encoding GetProviderEncoding(CmdletProvider provider, FileEncoding encoding)
+        {
+            Encoding resolvedEncoding = GetDefaultEncoding();
+            FileEncoding encodingPreference = GetEncodingPreference(provider.SessionState);
+            // If the encoding isn't set, but is available as $PSDefaultFileEncoding, use that
+            // It the encoding is set use that, otherwise return the default encoding
+            if ( encoding == FileEncoding.Unspecified && encodingPreference != FileEncoding.Unspecified )
+            {
+                resolvedEncoding = GetEncoding(encodingPreference);
+            }
+            else if ( encoding != FileEncoding.Unspecified )
+            {
+                resolvedEncoding = GetEncoding(encoding);
+            }
+            return resolvedEncoding;
+        }
+
+        // This is the way the encoding is implemented in PowerShell 5 and earlier.
+        // If the user sets the default encoding to WindowsLegacy, we will
+        // be able to encode for that
+        internal static Dictionary<String, Encoding> legacyEncodingMap =
+            new Dictionary<string, Encoding>(StringComparer.OrdinalIgnoreCase)
+            {
+                { "microsoft.powershell.commands.addcontentcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.exportclixmlcommand", Encoding.Unicode },
+                { "microsoft.powershell.commands.exportcsvcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.exportpssessioncommand", Encoding.UTF8 }, // with BOM
+                { "microsoft.powershell.commands.formathex", Encoding.ASCII },
+                { "microsoft.powershell.commands.newmodulemanifestcommand", Encoding.Unicode },
+                { "microsoft.powershell.commands.getcontentcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.importcsvcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.outfilecommand", Encoding.Unicode }, // This includes redirection
+                { "microsoft.powershell.commands.setcontentcommand", Encoding.ASCII },
+                // Providers are handled here
+                { "microsoft.powershell.commands.filesystemprovider", Encoding.ASCII },
+            };
```

Comment:
Unfortunately, that doesn't appear to be correct.

As for the _writing_ cmdlets:

`Add-Content` and `Set-Content` - despite what the documentation claims - have always used "ANSI" encoding, not ASCII - see https://github.com/PowerShell/PowerShell-Docs/issues/1483.

What about `Send-MailMessage`? The help topics that ASCII encoding is the default.

`New-Item -Type File -Value` currently creates BOM-less(!) UTF-8.

As an aside: While `Export-Csv` without `-Append` indeed creates ASCII files, `Export-Csv -Append` currently blindly appends _UTF-8_ if the existing file's encoding is any of ASCII/UTF-8/"ANSI", but correctly matches UTF-16LE and UTF-16BE.

---

As for the _reading_ cmdlets:

`Get-Content` currently defaults to "ANSI" in the absence of a BOM, as does `Import-PowerShellDataFile`.

`Import-Csv` currently actually assumes UTF-8 in the absence of a BOM, as do `Import-CliXml` and `Select-String`